File: backend/app/services/company_profile_service.py
"""Company profile business logic."""

from __future__ import annotations

from app.models.company_profile import CompanyProfile
from app.repositories.company_profile_repository import CompanyProfileRepository


class CompanyProfileService:
    """Manage the singleton company profile record."""

    def __init__(self, company_profiles: CompanyProfileRepository) -> None:
        self._company_profiles = company_profiles

    async def get_profile(self) -> CompanyProfile:
        return await self._company_profiles.get_or_create()

    async def update_profile(
        self,
        *,
        company_name: str,
        address: str,
        area: str,
        city: str,
        pin: str,
        email: str,
        phone: str,
        gstin: str,
    ) -> CompanyProfile:
        profile = await self._company_profiles.get_or_create()
        return await self._company_profiles.update(
            profile,
            company_name=company_name,
            address=address,
            area=area,
            city=city,
            pin=pin,
            email=email,
            phone=phone,
            gstin=gstin,
        )

File: frontend/src/services/companyProfileService.ts
import { apiClient } from "@/services/apiClient";
import type {
  CompanyProfileResponse,
  UpdateCompanyProfileRequest,
} from "@/types/api";

export type CompanyProfile = {
  id: number;
  companyName: string;
  address: string;
  area: string;
  city: string;
  pin: string;
  email: string;
  phone: string;
  gstin: string;
  createdAt: string;
  updatedAt: string;
};

function mapCompanyProfile(data: CompanyProfileResponse): CompanyProfile {
  return {
    id: data.id,
    companyName: data.company_name,
    address: data.address,
    area: data.area,
    city: data.city,
    pin: data.pin,
    email: data.email,
    phone: data.phone,
    gstin: data.gstin,
    createdAt: data.created_at,
    updatedAt: data.updated_at,
  };
}

export const companyProfileService = {
  async getProfile(): Promise<CompanyProfile> {
    const { data } = await apiClient.get<CompanyProfileResponse>("/company-profile");
    return mapCompanyProfile(data);
  },

  async updateProfile(
    payload: UpdateCompanyProfileRequest,
  ): Promise<CompanyProfile> {
    const { data } = await apiClient.patch<CompanyProfileResponse>(
      "/company-profile",
      payload,
    );
    return mapCompanyProfile(data);
  },
};
